src/AppInstallerCLICore/Workflows/UninstallFlow.cpp

```diff
@@ -0,0 +1,182 @@
+using namespace AppInstaller::Repository;
+
+namespace AppInstaller::CLI::Workflow
+{
+    namespace
+    {
+        std::optional<DWORD> InvokeCreateProcess(const std::string& command, IProgressCallback& progress)
+        {
+            std::wstring commandUtf16 = Utility::ConvertToUTF16(command);
+
+            // Parse the command string as application and command line for CreateProcess
+            wil::unique_cotaskmem_string app = nullptr;
+            wil::unique_cotaskmem_string commandLine = nullptr;
+            THROW_IF_FAILED(SHEvaluateSystemCommandTemplate(commandUtf16.c_str(), &app, &commandLine, NULL));
+
+            STARTUPINFOW startupInfo = { sizeof(startupInfo) };
+            wil::unique_process_information processInfo = {};
+            BOOL success = CreateProcessW(app.get(), commandLine.get(), NULL, NULL, FALSE, 0, NULL, NULL, &startupInfo, &processInfo);
+
+            if (!success)
+            {
+                return GetLastError();
+            }
+
+            // Wait for uninstallation to finish
+            while (!progress.IsCancelled())
+            {
+                DWORD waitResult = WaitForSingleObject(processInfo.hProcess, 250);
+                if (waitResult == WAIT_OBJECT_0)
+                {
+                    break;
+                }
+
+                if (waitResult != WAIT_TIMEOUT)
+                {
+                    THROW_LAST_ERROR_MSG("Unexpected WaitForSingleObjectResult: %d", waitResult);
+                }
+            }
+
+            if (progress.IsCancelled())
+            {
+                return {};
+            }
+            else
+            {
+                DWORD exitCode = 0;
+                GetExitCodeProcess(processInfo.hProcess, &exitCode);
+                return exitCode;
+            }
+        }
+    }
+
+    void GetUninstallInfo(Execution::Context& context)
+    {
+        auto installedPackageVersion = context.Get<Execution::Data::InstalledPackageVersion>();
+        const std::string installedTypeString = installedPackageVersion->GetMetadata()[PackageVersionMetadata::InstalledType];
+        switch (ManifestInstaller::ConvertToInstallerTypeEnum(installedTypeString))
+        {
+        case ManifestInstaller::InstallerTypeEnum::Exe:
+        case ManifestInstaller::InstallerTypeEnum::Burn:
+        case ManifestInstaller::InstallerTypeEnum::Inno:
+        case ManifestInstaller::InstallerTypeEnum::Msi:
+        case ManifestInstaller::InstallerTypeEnum::Nullsoft:
+        case ManifestInstaller::InstallerTypeEnum::Wix:
+        {
+            IPackageVersion::Metadata packageMetadata = installedPackageVersion->GetMetadata();
+            IPackageVersion::Metadata::iterator itr = packageMetadata.find(PackageVersionMetadata::StandardUninstallCommand);
+
+            if (itr == packageMetadata.end() || context.Args.Contains(Execution::Args::Type::Silent))
+            {
+                itr = packageMetadata.find(PackageVersionMetadata::SilentUninstallCommand);
```

Comment:
>itr = packageMetadata.find(PackageVersionMetadata::SilentUninstallCommand); [](start = 16, length = 75)

I think if --silent is passed in and there exists StndardUninstallString, this will lead to ERROR_NO_UNINSTALL_INFO_FOUND? Which is not what John wanted above?

Reply:
Agreed, this should get an another iterator and only assign to `itr` if it is valid.

---
In reply to: [536449750](https://github.com/Microsoft/winget-cli/pull/659#discussion_r536449750) [](ancestors = 536449750)